heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java

```diff
@@ -0,0 +1,419 @@
+ * size.
+ * <p>
+ * 3.The user provides the value of com.twitter.heron.api.Config.TOPOLOGY_ADJUST_NUMBER_CONTAINERS.
+ * In case the value is false and the algorithm fails to produce a packing plan
+ * that satisfies the resource requirements of all
+ * the components in the given number of containers, a null packing
+ * will be returned.
+ * If the value is true, the algorithm will automatically adjust the number of containers
+ * so that the resource requirements of all the components are satisfied.
+ * The default value of the parameter is false.
+ * <p>
+ * 4. The user provides a percentage of each container size that will be added to the resources
+ * allocated by the container through the com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_PADDING_PERCENTAGE
+ * If the parameter is not specified then a default value of 10 is used (10% of the container size)
+ * <p>
+ * 5. The ram required for one instance is calculated as:
+ * value in com.twitter.heron.api.Config.TOPOLOGY_COMPONENT_RAMMAP if exists, otherwise,
+ * the default ram value for one instance.
+ * <p>
+ * 6. The cpu required for one instance is calculated as the default cpu value for one instance.
+ * <p>
+ * 7. The disk required for one instance is calculated as the default disk value for one instance.
+ * <p>
+ * 8. The ram required for a container is calculated as:
+ * (ram for instances in container) + (paddingPercentage * ram for instances in container)
+ * <p>
+ * 9. The cpu required for a container is calculated as:
+ * (cpu for instances in container) + (paddingPercentage * cpu for instances in container)
+ * <p>
+ * 10. The disk required for a container is calculated as:
+ * (disk for instances in container) + ((paddingPercentage * disk for instances in container)
+ * <p>
+ * 11. The size of resources required by the whole topology is equal to
+ * ((# containers used by ResourceCompliantRR)* size of each container
+ * + size of a container that includes one instance with default resource requirements).
+ * We add some resources to consider the Heron internal container (i.e. the one containing Scheduler
+ * and TMaster).
+ * <p>
+ * 12. The pack() return null if PackingPlan fails to pass the safe check, for instance,
+ * the size of ram for an instance is less than the minimal required value or .
+ */
+public class ResourceCompliantRRPacking implements IPacking {
+
+  public static final long MIN_RAM_PER_INSTANCE = 192L * Constants.MB;
+  public static final int DEFAULT_CONTAINER_PADDING_PERCENTAGE = 10;
+
+  private static final Logger LOG = Logger.getLogger(ResourceCompliantRRPacking.class.getName());
+  protected TopologyAPI.Topology topology;
+
+  protected long instanceRamDefault;
+  protected double instanceCpuDefault;
+  protected long instanceDiskDefault;
+  protected long maxContainerRam;
+  protected double maxContainerCpu;
+  protected long maxContainerDisk;
+  protected int numContainer;
+  protected int noAdjustments;
```

Comment:
For consistency and clarity should change these to numContainers and numAdjustments. noAdjustments sounds like a boolean.